gluten-core/src/main/scala/io/glutenproject/extension/ColumnarOverrides.scala

```diff
@@ -502,26 +506,87 @@ case class ColumnarOverrideRules(session: SparkSession)
       overridden
     }
 
+  def checkColumnarToRow(plan: SparkPlan): Unit = {
+    plan match {
+      case _: ColumnarToRowExec =>
+        fallbacks = fallbacks + 1
+      case _ =>
+    }
+    plan.children.map(plan => checkColumnarToRow(plan))
+  }
+
+  def fallbackWholeStage(plan: SparkPlan): Boolean = {
+    if (wholeStageFallbackThreshold == -1) {
+      return false
+    }
+    fallbacks = 0
+    checkColumnarToRow(plan)
+    if (fallbacks >= wholeStageFallbackThreshold) {
+      true
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Ported from ApplyColumnarRulesAndInsertTransitions of Spark.
+   * Inserts an transition to columnar formatted data.
+   */
+  private def insertRowToColumnar(plan: SparkPlan): SparkPlan = {
+    if (!plan.supportsColumnar) {
+      // The tree feels kind of backwards
+      // Columnar Processing will start here, so transition from row to columnar
+      RowToColumnarExec(insertTransitions(plan, outputsColumnar = false))
+    } else if (!plan.isInstanceOf[RowToColumnarTransition]) {
+      plan.withNewChildren(plan.children.map(insertRowToColumnar))
+    } else {
+      plan
+    }
+  }
+
+  /**
+   * Ported from ApplyColumnarRulesAndInsertTransitions of Spark.
+   * Inserts RowToColumnarExecs and ColumnarToRowExecs where needed.
+   */
+  private def insertTransitions(plan: SparkPlan, outputsColumnar: Boolean): SparkPlan = {
+    if (outputsColumnar) {
+      insertRowToColumnar(plan)
+    } else if (plan.supportsColumnar) {
+      // `outputsColumnar` is false but the plan outputs columnar format, so add a
+      // to-row transition here.
+      ColumnarToRowExec(insertRowToColumnar(plan))
+    } else if (!plan.isInstanceOf[ColumnarToRowTransition]) {
+      plan.withNewChildren(plan.children.map(insertTransitions(_, outputsColumnar = false)))
+    } else {
+      plan
+    }
+  }
+
   override def postColumnarTransitions: Rule[SparkPlan] = plan => PhysicalPlanSelector.
     maybe(session, plan) {
-      logOnLevel(
-        transformPlanLogLevel,
-        s"postColumnarTransitions preOverriden plan:\n${plan.toString}")
-      var overridden: SparkPlan = plan
-      val startTime = System.nanoTime()
-      postOverrides(supportAdaptive).foreach { r =>
-        overridden = r(session)(overridden)
-        planChangeLogger.logRule(r(session).ruleName, plan, overridden)
+      if (supportAdaptive && fallbackWholeStage(plan)) {
```

Comment:
Will whole_stage_fall_back mechanism work for AQE dsiabled scenarios?That will make UT more convenient. Just like what WholeStageCodegenSuite does: with DisableAdaptiveExecutionSuite

Reply:
Thanks for your comment!
Our initial purpose is to support fallback for a whole query stage. If AQE config is on (generally turned on by users, right?) and the given plan is supported by adaptive execution, the query stage's plan will go through our injected columnar rule, which makes one stage by one stage fallback handling easy (see this PR). Otherwise, we would need to cautiously check each stage boundaries, which is a bit complex.
If AQE config is off, we can easily realize a fallback strategy with very few code changes here, but for the entire query plan, not for a stage.

We are not sure which is a better strategy: fall back a stage or an entire query plan. Maybe, it varies from case to case. You can help do some tests on your workloads. Then, we can do some changes accordingly.

Reply:
Is it considered that when the fallback count meets the threshold and then will fallback all the remaining plan, but not the already transformed ones ?

Reply:
> Is it considered that when the fallback count meets the threshold and then will fallback all the remaining plan, but not the already transformed ones ?

@zzcclp Thanks for your comment! In this patch, the fallback is a kind of per-stage fallback. And each stage's fallback handling is independent. We are considering to support your mentioned strategy in a follow-up PR, i.e., once the threshold is met, fall back all the remaining plans even though they are in another stage.